### DISCUSSION TITLE
How to configure custom kernels

### DISCUSSION THREAD

**Discussion by stefanseefeld** [nebari-dev/nebari#2553] [Q&A]
Hello,
I'm looking for information about setting up custom jupyter kernels with nebari. In particular, I want to configure my notebooks to run in custom docker containers, and thus need a way to pull the corresponding docker images into nebari so they can be started on demand. I was able to set this up in a TLJH instance, but am unsure of how to do the same with nebari. Any pointers would be highly appreciated. Thanks !

**Comment by Adam-D-Lewis** [comment]
Hi @stefanseefeld, are there any docs on the littlest Jupyter hub that explain what you're describing?  I'm Nebari, building your environments in conda store is definitely the most supported, but I have seen some deployments with conda environments in the docker image.  I'll try to provide some more info later.

Can you further describe why building your environment in conda store doesn't work for your use case?  Maybe there's a feature we could add to conda store.

**Reply by stefanseefeld** [reply]
I can unfortunately no longer find the docs I was using; it may actually have come from some external place, such as stackoverflow...

Our setup is this: In our projects we are mostly working with C++ (for deployment), but using many Python bindings (using Boost.Python) to establish an R&D platform. Most of our development is containerized, i.e. we have a few "base images", as well as a few "dev images" for the various projects, which contain the installed C++ libs together with their respective Python bindings.

On top of all this we use jupyter notebooks to prototype, document, and formally specify various algorithms. These notebooks are expected to be run within the dev images I described above. I'm now seeking a way for us to publish these notebooks, to make them accessible (for reading as well as execution) to a wider audience, i.e. not only software engineers, but also system engineers, domain experts, and other stakeholders.

I was able to pull the necessary dev images into the VM running TLJH, and then define corresponding kernels that would execute them. This worked fine, but I got stuck when trying to use CDS Dashboards to share them. I'm now trying to reproduce the same with Nebari, using JHub App Launcher.

**Reply by Adam-D-Lewis** [reply]
Thanks for the additional info.  I'd say we can see some benefit in supporting the features you need, but there is no current work or concrete plan to add them at the moment.  Talking with @krassowski, it looks like Nebari would need a refactor to use https://jupyter-enterprise-gateway.readthedocs.io/en/latest/ or https://github.com/jupyter-server/gateway_provisioners to support launching notebooks in specific docker images.  If the possibility of funding the intergration of those features exists, we would certainly be interested.

However, as it stands, all notebooks run within a single docker image.  You can customize that image however and run your own image.  To do so, you would add the following to your docker config.
```
default_images:
  jupyterhub: quay.io/nebari/nebari-jupyterhub:2024.6.1rc3
  jupyterlab: quay.io/nebari/nebari-jupyterlab:2024.6.1rc3
```
The docker images are built in https://github.com/nebari-dev/nebari-docker-images repo.  You could customize the jupyterlab image potentially adding new jupyter kernels, but at the moment it's not well tested.  I have seen a deployment doing that, but some features of Nebari weren't working (e.g. [Jupyter Scheduler](https://www.nebari.dev/docs/tutorials/jupyter-scheduler/)) due to it not being well supported.

**Reply by Adam-D-Lewis** [reply]
There is another idea that might work.  A docker in docker option **might** be possible as well, but it's really just an idea at this point, and I mention it just to record the thought.  We'd need to add docker to the jupyterlab image, then create a kernel like the following:
```
{
  "argv": [
    "docker",
    "run",
    "--rm",
    "-i",
    "-v", "{connection_file}:/connection_file",
    "your-docker-image:tag",
    "python",
    "-m",
    "ipykernel_launcher",
    "-f",
    "/connection_file"
  ],
  "display_name": "Docker Python",
  "language": "python"
}
```
I think the inner docker container would also need to run in  privileged mode.

**Reply by stefanseefeld** [reply]
Yes, the above kernel config is exactly how I defined my own docker-based kernels. That worked fine, at least in TLJH where everything was local. I don't know whether or how this would work on a cluster.

In fact, the main reason I'm exploring Nebari is because of the JHub App loader. Could I use that outside of Nebari, e.g. within TLJH ?

**Reply by Adam-D-Lewis** [reply]
JHub apps is intended to be usable on other Jupyterhub deployments besides Nebari, but I'm not sure how much testing has gone into testing it on TLJH for example.  @aktech could likely comment more on that.

**Reply by dharhas** [reply]
jhub-apps should work with any JupyterHub deployment with the caveat that the default theme has Nebari colors. If you try it out and get it working we'd love any documentation pr's if there is anything specific you encounter when using TLJH.

**Reply by stefanseefeld** [reply]
Thanks, that's great to hear. We are still having some issues setting things up (see https://github.com/nebari-dev/jhub-apps/issues/409), but will hopefully sort this out quickly.

**Reply by aktech** [reply]
Hey @stefanseefeld 👋 

I responded to that issue, earlier today. Jhub-apps should surely work on TLJH, I deployed it in the pas several times while developing. There might be some nitpicks required, which I might be able to help you with.
